galley/pkg/crd/synchronizer.go

```diff
@@ -0,0 +1,364 @@
+
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Synchronizer is a Kubernetes controller that keeps a synchronized copy of an source set of CRDs in
+// destination APIGroup/Version(s). The mapping from source to destination is specified by a mapping object.
+// A passed-in listener can be used to detect when target CRDs are added/deleted.
+type Synchronizer struct {
+
+	// mutex for controlling the start/stop behavior.
+	stateLock sync.Mutex
+
+	// mapping from the source APIGroup/Versions to destination APIGroup/Versions.
+	mapping Mapping
+
+	// The resync period to use when initializing an informer.
+	resyncPeriod time.Duration
+
+	// The standard Kubernetes interface for accessing CRDs.
+	crdi v1beta1.CustomResourceDefinitionInterface
+
+	// stopCh is used to quiesce the background activity during shutdown.
+	stopCh chan struct{}
+
+	// queue is used to schedule work in a single background go-process.
+	queue workqueue.RateLimitingInterface
+
+	// SharedIndexInformer for watching/caching CRDs.
+	informer cache.SharedIndexInformer
+
+	// Optional listener that receives notifications when destination CRDs get created/deleted.
+	listener SyncListener
+
+	// event hook that gets called after every completion of event processing loop. Useful for testing.
+	eventHook eventHookFn
+}
+
+// SyncListener defines an interface for a listener that will get notified as destination Crds are added/removed.
+type SyncListener struct {
+	OnDestinationAdded   func(name string, kind string, listKind string)
+	OnDestinationRemoved func(name string)
+}
+
+// NewSynchronizer returns a new instance of a Synchronizer. The returned Synchronizer is not started: this
+// needs to be done explicitly using start()/Stop() methods.
+func NewSynchronizer(config *rest.Config, mapping Mapping, resyncPeriod time.Duration, listener SyncListener) (s *Synchronizer, err error) {
+	return newSynchronizer(config, mapping, resyncPeriod, listener, nil, getCustomResourceDefinitionsInterface)
+}
+
+type eventHookFn func(e interface{})
+
+// NewSynchronizer returns a new instance of a Synchronizer. The returned Synchronizer is not started: this
+// needs to be done explicitly using start()/Stop() methods.
+func newSynchronizer(config *rest.Config, mapping Mapping, resyncPeriod time.Duration, listener SyncListener,
+	eventHook eventHookFn, crdiFn getCrdiFn) (s *Synchronizer, err error) {
+
+	var crdi v1beta1.CustomResourceDefinitionInterface
+	if crdi, err = crdiFn(config); err != nil {
+		return nil, err
+	}
+
+	s = &Synchronizer{
+		mapping:      mapping,
+		resyncPeriod: resyncPeriod,
+		listener:     listener,
+		crdi:         crdi,
+		eventHook:    eventHook,
+	}
+
+	return
+}
```

Comment:
Similar comment re: named return parameters. IMO, the following is more concise without the extra forward declarations a. la C89.

```golang
crdi, err := crdiFn(config)
if err != nil {
	return nil, err
}

return &Synchronizer{
	mapping:      mapping,
	resyncPeriod: resyncPeriod,
	listener:     listener,
	crdi:         crdi,
	eventHook:    eventHook,
}, nil
```